### PR TITLE
fix(bigquery): Fixes parsing of null type values returned from BQ server

### PR DESCRIPTION
**Pull request by jsrinnn**
This PR includes the following changes for json parsing related to null values, based on what is returned from BQ Server API

1. BQ query response can contain null values for row/column data as shown below

```
"rows": [
    {
      "f": [
        {
          "v": null
        },
        {
          "v": "Test String 6"
        },
        {
          "v": null
        }
      ]
    },
```
Due to the above json `get_to` method crashes with a type exception . The fix checks for null value before calling `get_to` in json_utils.h

2) From the client perspective, we need to be able to distingush an empty string from a null value so that client api can work properly hence an additional `is_null` member has been added to `ColumnData` struct.

3) Unit test changes for all of the above.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14384)

### CODE REVIEW

File: google/cloud/bigquery/v2/minimal/internal/common_v2_resources.h
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/common_v2_resources.h:319]: I think it would be more idiomatic to use `absl::optional<std::string>` to represent this.  At the very least it would avoid awkward situations like `value == "foo"` and `is_null == true`.

File: google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc:637]: If `c.value` was `absl::optional<std::string>` you could just use overloading on `SafeGetTo`?

**Review by coryan — COMMENTED** [review]
Please consider if `absl::optional<std::string>` is a better representation. If you have a good reason to prefer a separate `is_null` please let me know and  I can review again.

**Review by jsrinnn — COMMENTED** [review]
So the main reason for the additional parameter is minimum disruption to the client code. Otherwise modifying `std::string` to absl::optional will break the client code. In the past when this has happened we had to do the following for unblocking

1) Modify the client code to look at a non-stable version of google-cloud-cpp (i.e main branch)

2) Make the Client code changes 

3) Wait till fix is available in google-cloud-cpp is stable release

4) Revert the client code to a stable-version of google-cloud-cpp.

I am mainly trying to avoid the above.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14384#-:-O0Piwv60Wc8y9n0jDzK:b-djz5vz)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @coryan)

___
*[`google/cloud/bigquery/v2/minimal/internal/common_v2_resources.h` line 319 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14384#gh-1657504971:-O0Pjgn246mRbZYZUyt3:b4ioyfb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/705117cb2cbcb22c42aebf41e90918acd840a7fb/google/cloud/bigquery/v2/minimal/internal/common_v2_resources.h#L319)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think it would be more idiomatic to use `absl::optional<std::string>` to represent this.  At the very least it would avoid awkward situations like `value == "foo"` and `is_null == true`.
</blockquote></details>

Yeah...I thought about that but modifying the return value breaks our client code and its a lot of back and forth to get everything working as mentioned below

So the main reason for the additional parameter is minimum disruption to the client code. Otherwise modifying `std::string` to absl::optional will break the client code. In the past when this has happened we had to do the following for unblocking

1) Modify the client code to look at a non-stable version of google-cloud-cpp (i.e main branch)

2) Make the Client code changes 

3) Wait till fix is available in google-cloud-cpp is stable release

4) Revert the client code to a stable-version of google-cloud-cpp.

I am mainly trying to avoid the above.
___
*[`google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc` line 637 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14384#gh-1657512650:-O0PjxpCCFqMRWMHaed7:b-6i9yit) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/705117cb2cbcb22c42aebf41e90918acd840a7fb/google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc#L637)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

If `c.value` was `absl::optional<std::string>` you could just use overloading on `SafeGetTo`?
</blockquote></details>

Please see my comment above for why I am not modifying the return value.

**Comment by coryan** [comment]
If we go with the `is_null` approach: how will you clean the code up?

**Review by jsrinnn — COMMENTED** [review]
Couple of options here:

1) I'll be OOO fairly soon so I didn't want the team to be blocked hence the above approach seemed to be the least disruptive. I can merge this code now and then come back and do the cleanup of using `absl::optional`.

2) Someone from my team can patch this PR and do the `absl::optional` approach.

I am leaning towards (1) so as not to add additional load on the team and since this is an internal library. But I can understand as code owners if you think otherwise.

Let me know what you think.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14384#-:-O0PnYfK2oBc8SBzZ4qA:b-yxl6p2)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @coryan)

**Review by coryan — APPROVED** [review]
Please file a bug for the cleanup and then we can merge.

> Reviewed 6 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14384#-:-O0PrlrC1FAWZboRQ3t7:b-x1v39w)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @jsrinnn)

**Review by jsrinnn — COMMENTED** [review]
Sounds good! Filed a tracking bug(#14387) and added a TODO in the code.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14384#-:-O0Q0cWfBWVl9SpUM2ZT:br8gp77)* status: 5 of 6 files reviewed, all discussions resolved (waiting on @coryan)

**Pull request merged**
